text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,626 @@
+the "Create" or "Publish" span for the message. In addition, if it is possible
+the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
+"Receive" span.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
+MAY link to the "Create" or "Publish" span of the message.
+
+## Open issues
+
+Fully integrating the changes proposed in this document into messaging semantic
+conventions requires some additions and clarifications in the specification,
+which are listed in this section.
+
+* [open-telemetry/opentelemetry-specification#454](https://github.com/open-telemetry/opentelemetry-specification/issues/454)
+  To instrument push-based "Receive" operations as described in this document,
+  it is necessary to add links to spans after those spans were created. The
+  reason for this is, that not all messages are present at the start of a
+  "Receive" operations, so links to related contexts cannot be added at the start
+  of the span.
+* [open-telemetry/opentelemetry-specification#2176](https://github.com/open-telemetry/opentelemetry-specification/issues/2176)
+  When consuming a message with no attached creation context as part of a
+  batch, it would still be useful to capture related message-specific
+  attributes as part of a link which points to an invalid context. However,
+  according to the specification links pointing to an invalid context may be
+  ignored. For consistently reporting message-specific attributes on links, links
+  to invalid contexts should be allowed and supported.
+* [open-telemetry/opentelemetry-specification#3172](https://github.com/open-telemetry/opentelemetry-specification/issues/3172)
+  Currently the specification is unclear about whether relationships between
+  producer and consumer spans can be modelled via links, the wording suggests
+  that it should be a parent/child relationship. The wording in the specification
+  needs to make it clear that this can be a link too.
+
+## Examples
+
+This section contains a list of examples illustrating the use of the
+conventions outlined above. Green boxes denote spans that are required to
+exist in order to conform to those conventions. Other boxes denote spans that
+are not required and covered by the conventions, but are hopefully helpful in
+understanding how messaging spans can be integrated into an overall trace flow.
+Solid arrows denote parent/child relationships, dotted arrows denote link
+relationships.
+
+### Push-based scenarios
+
+A producer creates and publishes a single message, the single message is delivered to a consumer:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
```

Comment:
Would you say that the the link arrow should be `Producer < --- Consumer` instead of `Producer -- > Consumer`? I just noticed it now.. I have hacked around mermaid (backwards links [doesn't seem to be supported ](https://github.com/mermaid-js/mermaid/issues/2629#issuecomment-1289228061)) and this code

```
flowchart TB;
  subgraph CONSUMER
  DM1[Deliver m1]
  end
  subgraph PRODUCER
  PM1[Publish m1]
  end
  PM1 --- DM1 .-> |link|PM1
  DM1 --- PM1
  PM1 --- DM1 --> |parent|PM1
  DM1 --- PM1

  classDef normal fill:green
  class PM1,DM1 normal
  linkStyle 0,2,3,5 stroke:none
  linkStyle 1,4 color:green,stroke:green
  ```
  
  Produces this output:
  
  
![image](https://user-images.githubusercontent.com/5938087/230388811-af255084-b66d-422b-b8df-55d5e6a19b0d.png)

I'm also thinking that on the examples where we have "parent", it would be nice to use `top-to-bottom` to emphasize the "parent/child" relationship. Also name the "parent" link.

Reply:
Let me know, I can help out with this as well (perhaps sending a PR to your branch)

Reply:
I don't have a strong opinion here, if other people too think it's clearer that way, I'd be happy to make the changes.

Personally, I think it's a bit confusing to reverse the direction on the parent and link edges, as edges then would point against the logical flow. For parent/child relationships, a directed edge currently means "is-parent-of", which makes sense in my opinion.

Reply:
With our current design, consumers add link _to_ "producers". Looking at it this way, I just find the current flow in the diagram of `producer --link--> consumer` a bit contradicting. If consumers are the ones adding the link I find it more intuitive that the "arrow" points to the producer instead `producer <--link-- consumer`.

Reply:
I opened a PR with the changes for this. https://github.com/pyohannes/oteps/pull/3. Some diagrams look more "messy" than they looked before. No way to improve the arranging of things in it so I'm not sure. Maybe leaving as-is is the best.

Reply:
I decided to not go ahead with this, as the diagrams look way worse. I did extract the `parent` label on the relationships and now is part of this PR.